core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -800,6 +835,103 @@ private void updateStageExecutions()
             });
         }
 
+        public SubPlan overridePartitionCountRecursively(SubPlan subPlan, int oldPartitionCount, int newPartitionCount)
+        {
+            PlanFragment planFragment = subPlan.getFragment();
+            if (stageExecutions.containsKey(getStageId(planFragment.getId()))) {
+                // already started, nothing to change for subPlan and its descendants
+                return subPlan;
+            }
+
+            if (planFragment.getPartitionCount().isEmpty()) {
+                // no partitioning, then no need to insert extra exchanges to sources
+                return new SubPlan(
+                        planFragment,
+                        subPlan.getChildren().stream().map(child -> overridePartitionCountRecursively(child, oldPartitionCount, newPartitionCount)).collect(toImmutableList()));
+            }
+
+            if (subPlan.getChildren().stream().noneMatch(child -> stageExecutions.containsKey(getStageId(child.getFragment().getId())))) {
+                // none of the source stages have started, then no need to insert extra exchanges to sources
+                return new SubPlan(
+                        planFragment.withPartitionCount(Optional.of(newPartitionCount)),
+                        subPlan.getChildren().stream().map(child -> overridePartitionCountRecursively(child, oldPartitionCount, newPartitionCount)).collect(toImmutableList()));
+            }
+
+            List<RemoteSourceNode> remoteSourceNodes = subPlan.getFragment().getRemoteSourceNodes();
+            ImmutableList.Builder<SubPlan> newSources = ImmutableList.builder();
+            // TODO: the below code inserts extra exchange as the parent of all sources,
+            // can be optimized to those that have started/completed, but directly change the partition count for those who haven't started
+            for (SubPlan source : subPlan.getChildren()) {
+                PlanFragment sourceFragment = source.getFragment();
+                RemoteSourceNode sourceRemoteSourceNode = getOnlyElement(remoteSourceNodes.stream()
+                        .filter(remoteSourceNode -> remoteSourceNode.getSourceFragmentIds().contains(sourceFragment.getId()))
+                        .iterator());
+                requireNonNull(sourceRemoteSourceNode, "sourceRemoteSourceNode is null");
+                RemoteSourceNode runtimeAdaptiveRemoteSourceNode = new RemoteSourceNode(
+                        getRuntimeAdaptivePlanNodeId(sourceRemoteSourceNode.getId()),
+                        sourceFragment.getId(),
+                        sourceFragment.getOutputPartitioningScheme().getOutputLayout(),
+                        sourceRemoteSourceNode.getOrderingScheme(),
+                        sourceRemoteSourceNode.getExchangeType(),
+                        sourceRemoteSourceNode.getRetryPolicy());
+                PlanFragment runtimeAdaptivePlanFragment = new PlanFragment(
```

Comment:
@sopel39 can you verify if we can always just insert extra stage which does the additional partitioning? Are we sure that `partitiningScheme` taken from `sourceFragment` will be able to operate on the output of the `sourceFragment`?

Reply:
@losipiuk I'm not sure what this code is about. Extra stage will always hurt performance. It is possible to maybe improve existing operators in source stage?

Reply:
@sopel39 : updated commit message. Basically we find performance hurts when number of partitions is large, and we see that 200 partitions will be enough for most of the cases for FTE. So small queries will just use 200 partitions, for huge queries if we see the size of intermediate stage is large, then we change the partition count (and therefore need to insert extra shuffle stages)